File: app/src/main/java/mx/dev/franco/automusictagfixer/UI/faq/FaqAdapter.java
package mx.dev.franco.automusictagfixer.UI.faq;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.List;

import mx.dev.franco.automusictagfixer.R;

/**
 * Created by franco on 21/03/18.
 */

public class FaqAdapter extends RecyclerView.Adapter<FaqAdapter.QuestionItemHolder> {
    private List<QuestionItem> mList;
    private OnItemClick mListener;
    private int mCurrentExpandedPosition = -1;

    public interface OnItemClick{
        void onItemClick(int position, View view);
    }

    public FaqAdapter(List list, OnItemClick listener){
        mList = list;
        mListener = listener;
    }

    @NonNull
    @Override
    public QuestionItemHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        View v = layoutInflater.inflate(R.layout.question_item, parent, false);
        return new QuestionItemHolder(v, mListener);
    }

    @Override
    public void onBindViewHolder(@NonNull QuestionItemHolder holder, int position) {
        QuestionItem questionItem = mList.get(position);
        holder.question.setText(questionItem.getQuestion());
        holder.answer.setText(questionItem.getAnswer());
        if(questionItem.isExpanded()) {
            holder.answer.setVisibility(View.VISIBLE);
        }
        else {
            holder.answer.setVisibility(View.GONE);
        }
    }

    public void setActivePosition(int position) {
        QuestionItem collapsedQuestionItem;
        QuestionItem expandedQuestionItem;
        if(mCurrentExpandedPosition != position) {
            if(mCurrentExpandedPosition == -1) {
                expandedQuestionItem = mList.get(position);
                expandedQuestionItem.setExpanded(true);
                notifyItemChanged(position);
            }
            else {
                collapsedQuestionItem = mList.get(mCurrentExpandedPosition);
                collapsedQuestionItem.setExpanded(false);
                notifyItemChanged(mCurrentExpandedPosition);
                expandedQuestionItem = mList.get(position);
                expandedQuestionItem.setExpanded(true);
                notifyItemChanged(position);
            }
            mCurrentExpandedPosition = position;
        }
        else {
            collapsedQuestionItem = mList.get(mCurrentExpandedPosition);
            collapsedQuestionItem.setExpanded(false);
            notifyItemChanged(mCurrentExpandedPosition);
            mCurrentExpandedPosition = -1;
        }
    }

    @Override
    public int getItemCount() {
        if(mList != null)
            return mList.size();
        return 0;
    }

    public static class QuestionItemHolder extends RecyclerView.ViewHolder implements View.OnClickListener{
        public TextView question;
        public TextView answer;
        public LinearLayout container;
        public OnItemClick listener;

        QuestionItemHolder(View root, OnItemClick listener){
            super(root);
            question = root.findViewById(R.id.question);
            answer = root.findViewById(R.id.answer);
            container = root.findViewById(R.id.container_question);
            this.listener = listener;
            root.setOnClickListener(this);
            //Enable animations on this layout
            /*container.getLayoutTransition()
                    .enableTransitionType(LayoutTransition.CHANGING);*/
        }

        @Override
        public void onClick(View v) {
            if(listener != null){
                listener.onItemClick(getAdapterPosition(), v);
            }
        }
    }

}

File: app/src/main/java/mx/dev/franco/automusictagfixer/ui/faq/FaqAdapter.kt
package mx.dev.franco.automusictagfixer.ui.faq

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import mx.dev.franco.automusictagfixer.R
import mx.dev.franco.automusictagfixer.ui.faq.FaqAdapter.QuestionItemHolder

/**
 * Created by franco on 21/03/18.
 */
class FaqAdapter(list: List<QuestionItem>?, listener: OnItemClick) :
    RecyclerView.Adapter<QuestionItemHolder>() {
    private val mList: List<QuestionItem>? = list
    private val mListener: OnItemClick = listener
    private var mCurrentExpandedPosition = -1

    interface OnItemClick {
        fun onItemClick(position: Int, view: View?)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): QuestionItemHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        val v = layoutInflater.inflate(R.layout.question_item, parent, false)
        return QuestionItemHolder(v, mListener)
    }

    override fun onBindViewHolder(holder: QuestionItemHolder, position: Int) {
        val questionItem = mList!![position]
        holder.question.text = questionItem.question
        holder.answer.text = questionItem.answer
        if (questionItem.isExpanded) {
            holder.answer.visibility = View.VISIBLE
        } else {
            holder.answer.visibility = View.GONE
        }
    }

    fun setActivePosition(position: Int) {
        val collapsedQuestionItem: QuestionItem
        val expandedQuestionItem: QuestionItem
        if (mCurrentExpandedPosition != position) {
            if (mCurrentExpandedPosition == -1) {
                expandedQuestionItem = mList!![position]
                expandedQuestionItem.isExpanded = true
                notifyItemChanged(position)
            } else {
                collapsedQuestionItem = mList!![mCurrentExpandedPosition]
                collapsedQuestionItem.isExpanded = false
                notifyItemChanged(mCurrentExpandedPosition)
                expandedQuestionItem = mList[position]
                expandedQuestionItem.isExpanded = true
                notifyItemChanged(position)
            }
            mCurrentExpandedPosition = position
        } else {
            collapsedQuestionItem = mList!![mCurrentExpandedPosition]
            collapsedQuestionItem.isExpanded = false
            notifyItemChanged(mCurrentExpandedPosition)
            mCurrentExpandedPosition = -1
        }
    }

    override fun getItemCount(): Int {
        return mList?.size ?: 0
    }

    class QuestionItemHolder internal constructor(root: View, var listener: OnItemClick?) :
        RecyclerView.ViewHolder(root), View.OnClickListener {
        var question: TextView = root.findViewById(R.id.question)
        var answer: TextView = root.findViewById(R.id.answer)
        var container: LinearLayout = root.findViewById(R.id.container_question)
        override fun onClick(v: View) {
            if (listener != null) {
                listener!!.onItemClick(adapterPosition, v)
            }
        }

        init {
            root.setOnClickListener(this)
        }
    }

}